### PR TITLE
8342650: Move getChars to DecimalDigits

### PR DESCRIPTION
**Pull request by wenshao**
Move getChars methods of StringLatin1 and StringUTF16 to DecimalDigits to reduce duplication

1. HexDigits and OctalDigits also include getCharsLatin1 and getCharsUTF16
2. Putting these two methods into DecimalDigits can avoid the need to expose them in JavaLangAccess
3. Eliminate duplicate code in BigDecimal

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8342650](https://bugs.openjdk.org/browse/JDK-8342650): Move getChars to DecimalDigits (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21593/head:pull/21593` \
`$ git checkout pull/21593`

Update a local copy of the PR: \
`$ git checkout pull/21593` \
`$ git pull https://git.openjdk.org/jdk.git pull/21593/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21593`

View PR using the GUI difftool: \
`$ git pr show -t 21593`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21593.diff">https://git.openjdk.org/jdk/pull/21593.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21593#issuecomment-2424714464)
</details>

### CODE REVIEW

File: src/java.base/share/classes/java/math/BigDecimal.java [hunk 1]
```diff
@@ -4284,18 +4197,24 @@ private String layoutChars(boolean sci) {
...
         }
 
-        StringBuilderHelper sbHelper = new StringBuilderHelper();
         char[] coeff;
         int offset;  // offset is the starting index for coeff array
         // Get the significand as an absolute value
         if (intCompact != INFLATED) {
-            offset = sbHelper.putIntCompact(Math.abs(intCompact));
-            coeff  = sbHelper.getCompactCharArray();
+            coeff = new char[19];
```
Comment (j3graham) [inline src/java.base/share/classes/java/math/BigDecimal.java:4216]: A possibility here would be to change `coeff` to be a String. The “else” branch already creates a string and has to additionally create a char array from it. If this is the only place where the `DecimalDigits.getChars(… char[])`, some extra code duplication could be removed. The `offset` could then also be removed

File: src/java.base/share/classes/jdk/internal/util/DecimalDigits.java [hunk 2]
```diff
@@ -136,4 +131,309 @@ public static int stringSize(long x) {
...
+     * Insert the 2-bytes integer into the buf as 2 decimal digit ASCII bytes,
+     * only least significant 16 bits of {@code v} are used.
+     * @param buf byte buffer to copy into
+     * @param charPos insert point
+     * @param v to convert
+     */
+    public static void putPairLatin1(byte[] buf, int charPos, int v) {
+        int packed = DIGITS[v];
+        putCharLatin1(buf, charPos, packed & 0xFF);
+        putCharLatin1(buf, charPos + 1, packed >> 8);
```
[the remark this answers is not in the document]
Comment (wenshao) [reply src/java.base/share/classes/jdk/internal/util/DecimalDigits.java:415]: Tests show that the current version of MergeStore also works

```
[TraceMergeStores] found:
    0-->  114  StoreB  === 80 91 113 95  [[ 18 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:24 (line 413)
    1-->   91  StoreB  === 80 7 90 51  [[ 114 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:13 (line 412)
[TraceMergeStores] truncated:
    0-->  114  StoreB  === 80 91 113 95  [[ 18 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:24 (line 413)
    1-->   91  StoreB  === 80 7 90 51  [[ 114 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:13 (line 412)
[TraceMergeStores]: Replace
  91  StoreB  === 80 7 90 51  [[ 114 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:13 (line 412)
 114  StoreB  === 80 91 113 95  [[ 18 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=5; unsafe  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:24 (line 413)
[TraceMergeStores]: with
  51  LoadS  === 33 7 49  [[ 91 85 95 70 118 ]]  @short[int:>=0] (java/lang/Cloneable,java/io/Serializable):exact+any *, idx=4; #short !jvms: DecimalDigits::putPairLatin1 @ bci:4 (line 411)
 118  StoreC  === 80 7 90 51  [[ ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; mismatched  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5;
[TraceMergeStores] MergePrimitiveStores::run:  118  StoreC  === 80 7 90 51  [[ 18 ]]  @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; mismatched  Memory: @byte[int:>=0] (java/lang/Cloneable,java/io/Serializable):NotNull:exact+any *, idx=5; !orig=[114] !jvms: DecimalDigits::putCharLatin1 @ bci:12 (line 430) DecimalDigits::putPairLatin1 @ bci:24 (line 413)
```

**Review by liach — APPROVED** [review]
Looks good. Tier 1-3 tests pass.

**Comment by wenshao** [comment]
This PR will improve the performance of Integer/Long.toString and StringBuilder.append(int/long) scenarios. This is because Unsafe.putByte is used to eliminate array bounds checks, and of course this elimination is safe.

In previous versions, in Integer/Long.toString and StringBuilder.append(int/long) scenarios, `-COMPACT_STRING` performed better than `+COMPACT_STRING`. This is because StringUTF16.getChars uses StringUTF16.putChar, which is similar to Unsafe.putChar, and there is no bounds check.

**Comment by TobiHartmann** [comment]
This caused a regression, see [JDK-8343925](https://bugs.openjdk.org/browse/JDK-8343925). You might want to consider a quick [backout](https://openjdk.org/guide/#backing-out-a-change).

**Comment by liach** [comment]
I think the fix is easy: Unsafe calls are using a bad offset which should be cast to long.  Will submit a PR, but since I cannot reproduce the original issue locally or on internal CI, need others review first.

**Pull request closed without merging**